balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,658 @@
+		if !b.addressList.increment() {
+			b.endFirstPass(scd.lastErr)
+			return
+		}
+		b.requestConnection()
+	case connectivity.Ready:
+		// Should never happen.
+		b.logger.Errorf("Requesting a connection even though we have a READY subconn")
+	}
+}
+
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) updateSubConnState(sd *scData, newState balancer.SubConnState) {
+	// Previously relevant subconns can still callback with state updates.
+	// To prevent pickers from returning these obsolete subconns, this logic
+	// is included to check if the current list of active subconns includes this
+	// subconn.
+	if activeSD, found := b.subConns.Get(sd.addr); !found || activeSD != sd {
+		return
+	}
+	if newState.ConnectivityState == connectivity.Shutdown {
+		return
+	}
+
+	if newState.ConnectivityState == connectivity.Ready {
+		b.shutdownRemaining(sd)
+		if !b.addressList.seekTo(sd.addr) {
+			// This should not fail as we should have only one subconn after
+			// entering READY. The subconn should be present in the addressList.
+			b.logger.Errorf("Address %q not found address list in  %v", sd.addr, b.addressList.addresses)
+			return
+		}
+		b.state = connectivity.Ready
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+		return
+	}
+
+	// If the LB policy is READY, and it receives a subchannel state change,
+	// it means that the READY subchannel has failed.
+	if b.state == connectivity.Ready && newState.ConnectivityState != connectivity.Ready {
+		// Once a transport fails, the balancer enters IDLE and starts from
+		// the first address when the picker is used.
+		b.state = connectivity.Idle
+		b.addressList.reset()
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker:            &idlePicker{exitIdle: b.ExitIdle},
+		})
+		return
+	}
+
+	if b.firstPass {
+		switch newState.ConnectivityState {
+		case connectivity.Connecting:
+			// The balancer can be in either IDLE, CONNECTING or TRANSIENT_FAILURE.
+			// If it's in TRANSIENT_FAILURE, stay in TRANSIENT_FAILURE until
+			// it's READY. See A62.
+			// If the balancer is already in CONNECTING, no update is needed.
+			if b.state == connectivity.Idle {
+				b.state = connectivity.Connecting
+				b.cc.UpdateState(balancer.State{
+					ConnectivityState: connectivity.Connecting,
+					Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+				})
+			}
+		case connectivity.TransientFailure:
+			sd.lastErr = newState.ConnectionError
+			// Since we're re-using common subconns while handling resolver updates,
+			// we could receive an out of turn TRANSIENT_FAILURE from a pass
+			// over the previous address list. We ignore such updates.
+
+			if curAddr := b.addressList.currentAddress(); !equalAddressIgnoringBalAttributes(&curAddr, &sd.addr) {
+				return
+			}
+			if b.addressList.increment() {
```

Comment:
I didn't add any code specifically for happy eyeballs since it would not be covered by any tests and would increase the scope of the PR.

> Should we make the code ready for this scenario since it's coming immediately next though?

What is the change you're suggesting here? IIUC, we should ignore any TFs that are not reported for the current address, even when happy eyeballs is implemented.

Reply:
> What is the change you're suggesting here? IIUC, we should ignore any TFs that are not reported for the current address, even when happy eyeballs is implemented.

We can't completely ignore it then.  Because we could have multiple connection attempts concurrently, we need to mark any connection attempt as having failed for the purposes of considering the first pass done.

I'm fine with leaving out things dealing with happy eyeballs, but we need to be sure it's well tested and there are some pretty interesting corner cases with the ordering of events.